### DISCUSSION TITLE
Is there any MicroPython libraries available for Industrial Protocols?

### DISCUSSION THREAD

**Discussion by Sai-Abishek** [micropython/micropython#9748] [Libraries & Drivers]
Hi there,
     I am planning to program ESP32 with MicroPython and looking for some MicroPython libraries that support Industrial protocols like EthernetIP, RS232 etc. I found some of the libraries like Pycomm and Pyserial. Now I need to know whether those same libraries can be used in MicroPython itself. Or Is there any other minimized version of the same? .

**Comment by karfas** [comment]
RS232 is a physical interface definition, no protocol. You get it e.g. with a MAX232 attached to an UART.
Ethernet/IP is a specification on top of TCP and UDP over standard ethernet.
You should search for the specific **protocols** required to communicate with your devices.

**Reply by Sai-Abishek** [reply]
> RS232 is a physical interface definition, no protocol. You get it e.g. with a MAX232 attached to an UART. Ethernet/IP is a specification on top of TCP and UDP over standard ethernet. You should search for the specific **protocols** required to communicate with your devices.

Hi Karfas,
    Thank you for the reply. I agree with what you say. I wrongly mentioned RS232 as a protocol and I am sorry for that. But in case of EthernetIP there is dedicated library named Pycomm available in python for communicating with Allen Bradley PLC .I want to do the same and need to communicate with Allen Bradley PLC using ESP32 . Now my question is, Is it possible to use that same Pycomm in  MicroPython or do I need to look for the specific protocols separately as you mentioned?. Sorry If my questions are very basic and stupid. I am new to this field and would like to learn more.

**Comment by andypiper** [comment]
I'd always suggest taking a look at [the Awesome MicroPython list](https://awesome-micropython.com/) to see what is there. There are a number of entries there for serial-related libraries. Beyond that, you'd probably need to be much more specific about the protocols you're looking to implement in your code.

**Reply by andypiper** [reply]
I think we posted at the same time, and now I see you mentioned the Allen Bradley PLC specifically. I don't see references to Logix or SLC on the list right now, but there's a decent chance that folks in this community who use MicroPython commercially may be aware of some options for you.

**Reply by Sai-Abishek** [reply]
@andypiper Sure I'll  take a look at it. Thank you

**Reply by andypiper** [reply]
Another source of knowledge might be to hop into the MicroPython Discord server to see whether folks there are aware of implementations that could get you jump-started.

**Reply by andypiper** [reply]
There is also [a related discussion](https://forum.micropython.org/viewtopic.php?f=2&t=8051&p=46431&hilit=plc#p45800) on the (archived) old MicroPython forum.

**Reply by Sai-Abishek** [reply]
> I think we posted at the same time, and now I see you mentioned the Allen Bradley PLC specifically. I don't see references to Logix or SLC on the list right now, but there's a decent chance that folks in this community who use MicroPython commercially may be aware of some options for you.

I gone through the link you sent. It seems to be excellent and there are many useful libraries available in the list. Anyway by the side I'll look for some options regarding this PLC communication. And as you mentioned I hope someone might be aware the thing I'm looking for. Thank You for the suggestion @andypiper

**Comment by sosi-deadeye** [comment]
You can look here for existing Micropython-Packages: https://awesome-micropython.com/
The only ready-to-use Protocols specially for Micropython for industrial automation I know are: CAN, mqtt, modbus

modbus can use TCP, UDP or TTY/RS232.
mqtt is only over TCP/IP
CAN require additional hardware (e.g. mcp2515)

Things that would be hard to implement:
- Profinet (requires membership)
- OPCUA (it's a monster)

You can also just use TCP/IP to exchange data.
UDP (connectionless) is a bit easier, if the order of received Packets doesn't matter.

Using the low-level networking stuff, requires understanding the data-types used in PLC and the byte-order.

**Reply by Sai-Abishek** [reply]
@sosi-deadeye Sure I'll look into it. Thanks for good pointers.
